service/mapper/fsm.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package mapper
+
+import (
+	"context"
+	"fmt"
+)
+
+type Edge struct {
+	check      CheckFunc
+	transition TransitionFunc
+}
+
+type FSM struct {
+	edges []Edge
+}
+
+func NewFSM() *FSM {
+
+	f := &FSM{}
+
+	return f
+}
+
+func (f *FSM) Add(check CheckFunc, transition TransitionFunc) {
+	edge := Edge{
+		check:      check,
+		transition: transition,
+	}
+	f.edges = append(f.edges, edge)
+}
+
+func (f *FSM) Run(s *State) error {
+TransitionLoop:
+	for {
+		for _, edge := range f.edges {
+			ok := edge.check(s)
+			if !ok {
+				continue
+			}
+			err := edge.transition(s)
+			if err != nil {
+				return fmt.Errorf("could not apply transition to state: %w", err)
+			}
+			continue TransitionLoop
+		}
+		return fmt.Errorf("could not find transition for state")
+	}
+}
```

Comment:
```suggestion
func (f *FSM) Run(s *State) error {
	for {
		for _, edge := range f.edges {
			ok := edge.check(s)
			if !ok {
				return fmt.Errorf("could not find transition for state")
			}
			err := edge.transition(s)
			if err != nil {
				return fmt.Errorf("could not apply transition to state: %w", err)
			}
			break
		}
	}
}
```

I feel like this is easier to follow. What do you guys think?

Reply:
2nd version looks better to me but it doesn't look like doing the same :) at least it returns error after the first `edge.check` failed which doesn't happen in the first version. May be I'm missing something though :)

Reply:
@evt is right, in your version, we error when the first check fails. We need to go through all checks to find the one that works, and only error if none of them do.